### PR TITLE
fix: task flow dynamic mapping with default_args

### LINKED ISSUE #33600: Dynamic tasks fail when default_args contains custom fields
### Apache Airflow version

Other Airflow 2 version (please specify below)

### What happened

Airflow version: **2.6.3**

Related to: https://github.com/apache/airflow/issues/29366

**DAG**

```python3
DEFAULT_TASK_ARGS = {
    "owner": "myteam",
    "start_date": "2023-08-18",
    "project_id": "myproject",
    "custom_arg": "foo",
}

@task
def foo() -> list[str]:
    return ["a", "b", "c"]

@task
def bar(val):
    logging.info(val)

with DAG(
    ...
    default_args=DEFAULT_TASK_ARGS,
) as dag:
    dynamic_tasks = bar.expand(val=foo())

    # Some more tasks that use `project_id` and `custom_arg`
    ...
```

If a DAG specifies **default_args** and the result of one of its tasks (**foo**) is used to expand another task (**bar**), expanded tasks fail at runtime with the following error:

```
[2023-08-21, 21:07:16 UTC] {taskinstance.py:1824} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.10/site-packages/airflow/models/taskinstance.py", line 1407, in _run_raw_task
    self._execute_task_with_callbacks(context, test_mode)
  File "/home/airflow/.local/lib/python3.10/site-packages/airflow/models/taskinstance.py", line 1531, in _execute_task_with_callbacks
    task_orig = self.render_templates(context=context)
  File "/home/airflow/.local/lib/python3.10/site-packages/airflow/models/taskinstance.py", line 2179, in render_templates
    original_task.render_template_fields(context)
  ...
  File "/home/airflow/.local/lib/python3.10/site-packages/airflow/models/baseoperator.py", line 429, in apply_defaults
    result = func(self, **kwargs, default_args=default_args)
  File "/home/airflow/.local/lib/python3.10/site-packages/airflow/models/baseoperator.py", line 788, in __init__
    raise AirflowException(
airflow.exceptions.AirflowException: Invalid arguments were passed to _PythonDecoratedOperator (task_id: bar__1). Invalid arguments were:
**kwargs: {'project_id': "myproject", 'custom_arg': "foo"}
```

Tried to add **kwargs** to the list of arguments, but got the same result

```python3
@task
def bar(val, **kwargs):
    logging.info(val)
```

Finally, tried calling **partial** before expanding to test if [this](https://github.com/apache/airflow/pull/29913) could fix it

```python3
dynamic_tasks = bar.partial().expand(val=foo())
```

### What you think should happen instead

Mapped tasks ignore non-generic default task arguments and run tasks accordingly.

### How to reproduce

Create a DAG similar to the one in the `What happened` section and run it.

### Operating System

Debian GNU/Linux 11 (bullseye)

### Versions of Apache Airflow Providers

apache-airflow-providers-airbyte==3.3.1
apache-airflow-providers-alibaba==2.4.1
apache-airflow-providers-amazon==8.2.0
apache-airflow-providers-apache-beam==5.1.1
apache-airflow-providers-apache-cassandra==3.2.1
apache-airflow-providers-apache-drill==2.4.1
apache-airflow-providers-apache-druid==3.4.1
apache-airflow-providers-apache-flink==1.1.1
apache-airflow-providers-apache-hdfs==4.1.0
apache-airflow-providers-apache-hive==6.1.1
apache-airflow-providers-apache-impala==1.1.1
apache-airflow-providers-apache-kylin==3.2.1
apache-airflow-providers-apache-livy==3.5.1
apache-airflow-providers-apache-pig==4.1.1
apache-airflow-providers-apache-pinot==4.1.1
apache-airflow-providers-apache-spark==4.1.1
apache-airflow-providers-apache-sqoop==3.2.1
apache-airflow-providers-arangodb==2.2.1
apache-airflow-providers-asana==2.2.1
apache-airflow-providers-atlassian-jira==2.1.1
apache-airflow-providers-celery==3.2.1
apache-airflow-providers-cloudant==3.2.1
apache-airflow-providers-cncf-kubernetes==7.4.1
apache-airflow-providers-common-sql==1.5.2
apache-airflow-providers-databricks==4.3.0
apache-airflow-providers-datadog==3.3.1
apache-airflow-providers-dbt-cloud==3.2.1
apache-airflow-providers-dingding==3.2.1
apache-airflow-providers-discord==3.3.0
apache-airflow-providers-docker==3.7.1
apache-airflow-providers-elasticsearch==4.5.1
apache-airflow-providers-exasol==4.2.1
apache-airflow-providers-facebook==3.2.1
apache-airflow-providers-ftp==3.4.2
apache-airflow-providers-github==2.3.1
apache-airflow-providers-google==10.5.0
apache-airflow-providers-grpc==3.2.1
apache-airflow-providers-hashicorp==3.4.1
apache-airflow-providers-http==4.4.2
apache-airflow-providers-imap==3.2.2
apache-airflow-providers-influxdb==2.2.1
apache-airflow-providers-jdbc==4.0.0
apache-airflow-providers-jenkins==3.3.1
apache-airflow-providers-microsoft-azure==6.1.2
apache-airflow-providers-microsoft-mssql==3.4.1
apache-airflow-providers-microsoft-psrp==2.3.1
apache-airflow-providers-microsoft-winrm==3.2.1
apache-airflow-providers-mongo==3.2.1
apache-airflow-providers-mysql==5.1.1
apache-airflow-providers-neo4j==3.3.1
apache-airflow-providers-odbc==4.0.0
apache-airflow-providers-openfaas==3.2.1
apache-airflow-providers-opsgenie==5.1.1
apache-airflow-providers-oracle==3.7.1
apache-airflow-providers-pagerduty==3.3.0
apache-airflow-providers-papermill==3.2.1
apache-airflow-providers-plexus==3.2.1
apache-airflow-providers-postgres==5.5.1
apache-airflow-providers-presto==5.1.1
apache-airflow-providers-qubole==3.4.1
apache-airflow-providers-redis==3.2.1
apache-airflow-providers-salesforce==5.4.1
apache-airflow-providers-samba==4.2.1
apache-airflow-providers-segment==3.2.1
apache-airflow-providers-sendgrid==3.2.1
apache-airflow-providers-sftp==4.3.1
apache-airflow-providers-singularity==3.2.1
apache-airflow-providers-slack==7.3.1
apache-airflow-providers-smtp==1.2.0
apache-airflow-providers-snowflake==4.2.0
apache-airflow-providers-sqlite==3.4.2
apache-airflow-providers-ssh==3.7.1
apache-airflow-providers-tableau==4.2.1
apache-airflow-providers-tabular==1.2.1
apache-airflow-providers-telegram==4.1.1
apache-airflow-providers-trino==5.1.1
apache-airflow-providers-vertica==3.4.1
apache-airflow-providers-zendesk==4.3.1

### Deployment

Other 3rd-party Helm chart

### Deployment details

Chart based on the official helm chart. Airflow running on Google Kubernetes Engine (GKE) using KubernetesExecutor.

### Anything else

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by phi-friday**
closes: #33600

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/decorators/base.py [hunk 1]
```diff
@@ -431,18 +431,67 @@ def _expand(self, expand_input: ExpandInput, *, strict: bool) -> XComArg:
...
-            is_setup=self.is_setup,
-            is_teardown=self.is_teardown,
-            on_failure_fail_dagrun=self.on_failure_fail_dagrun,
-        )
+        partial_kwargs: dict[str, Any] = {
+            "is_setup": self.is_setup,
+            "is_teardown": self.is_teardown,
+            "on_failure_fail_dagrun": self.on_failure_fail_dagrun,
+        }
+        # should be kept in sync with BaseOperatorMeta.partial().partial_kwargs.
```
Comment (uranusjr) [inline airflow/decorators/base.py]: Can we get these argument names with `inspect` instead?

Comment (phi-friday) [reply airflow/decorators/base.py]: Are you talking about `partial_keys`?
I tried it ```sig = inspect.signature(BaseOperator)```, and it makes a difference as much as 

```python
{'default_args',
 'do_xcom_push',
 'email_on_failure',
 'email_on_retry',
 'kwargs',
 'multiple_outputs',
 'params',
 'post_execute',
 'pre_execute',
 'task_concurrency'}
```

I think I should remove `default_args`, `kwargs`, `multiple_outputs`, and `params`, but other than that, can I use them as they are?

Comment (uranusjr) [reply airflow/decorators/base.py]: `do_xcom_push` is in the same boat as `multiple_outputs`. `task_concurrency` is deprecated so I don’t think we need to handle that (just ignore it). The rest I think can be used as-is.

File: airflow/decorators/base.py [hunk 2]
```diff
@@ -431,18 +431,29 @@ def _expand(self, expand_input: ExpandInput, *, strict: bool) -> XComArg:
...
+        }
+        base_signature = inspect.signature(BaseOperator)
+        ignore = {
+            "default_args",
+            "kwargs",
+            "do_xcom_push",
+            "multiple_outputs",
+            "params",
+            "task_concurrency",
+        }
```
Comment (uranusjr) [inline airflow/decorators/base.py]: Let’s add some comments to explain why each one is ignored.

**Review by uranusjr — COMMENTED** [review]
One nit, otherwise looks good to me

File: airflow/decorators/base.py (hunk 2 above)
Comment (phi-friday) [reply airflow/decorators/base.py]: First, as I mentioned earlier, here's why I thought we should remove `default_args`, `kwargs`, `multiple_outputs`, and `params`.
- `default_args`: This is the target we're working on right now.
- `kwargs`: A common name for a keyword argument.
- `multiple_outputs`: We use `self.multiple_outputs` in subsequent code.
- `params`: Already managed by `partial_params`.

and
- `task_concurrency`: is deprecated(will be replaced by `max_active_tis_per_dag`)
- `do_xcom_push`: I'm actually not sure about `do_xcom_push` yet.

Comment (uranusjr) [reply airflow/decorators/base.py]: Yeah I meant adding the information to the code base so readers don’t need to come back here for the context.

Comment (phi-friday) [reply airflow/decorators/base.py]: I'm still not quite sure about `do_xcom_push`, so if you can tell me what comments I should add to that, I'll add them to the commit.

Comment (phi-friday) [reply airflow/decorators/base.py]: I don't quite understand, but should I comment "in the same boat as `multiple_outputs`"?

Comment (phi-friday) [reply airflow/decorators/base.py]: I added it for now.

**Review by uranusjr — APPROVED** [review]
Good enough for me

**Comment by phi-friday** [comment]
After a recent rebase, I'm getting the following error, and I'm not sure why.

```shell
_________ TestOtelMetrics.test_timer_start_and_stop_manually_send_true _________
[gw2] linux -- Python 3.8.20 /usr/local/bin/python

self = <tests.core.test_otel_logger.TestOtelMetrics object at 0x7f0c14b7e370>
mock_time = <MagicMock name='perf_counter' id='139690224843072'>
name = 'test_stats_run'

    @mock.patch.object(time, "perf_counter", side_effect=[0.0, 3.14])
    def test_timer_start_and_stop_manually_send_true(self, mock_time, name):
        timer = self.stats.timer(name)
        timer.start()
        # Perform some task
        timer.stop(send=True)
    
        assert isinstance(timer.duration, float)
>       assert timer.duration == 3.14
E       assert 3140.0 == 3.14
E        +  where 3140.0 = <airflow.metrics.otel_logger._OtelTimer object at 0x7f0c2a6999a0>.duration

tests/core/test_otel_logger.py:327: AssertionError
```

**Comment by bitomukesh** [comment]
<div>
<h2>Changelist by Bito</h2>
    <p>This pull request implements the following key changes.</p>

<div>
<table width="100%">
<tbody>
  <tr>
    <th>Key Change</th>
    <th>Files Impacted</th>
  </tr>

<tr>
<td width="30%">
    <b>Bug Fix</b> - Fix Task Flow Dynamic Mapping with Default Arguments
</td>
<td width="70%">


<div>

-&nbsp;[`base.py`](https://github.com/apache/airflow/pull/41592/files#diff-3cbb8cc86a8ad7dbfed7f235680644a529dfa25ffdbcd71580a53f0e98c65cb5) - Updated handling of default arguments in task decorators
<div>

-&nbsp;[`test_mapped.py`](https://github.com/apache/airflow/pull/41592/files#diff-8c1514df75c893abef9fe5947769effafe8c247167bba8f031d99ce66a0cc6fb) - Added test for mapped task with arbitrary default arguments
</div>
</td>
</tr>

</tbody>
</table>
</div>
</div>

**Review by potiuk — APPROVED** [review]

**Pull request merged**